### PR TITLE
Adds a builder for building popup menu of a treeview item

### PR DESCRIPTION
**Pull request by naaando**
The reason for this refactoring is that the building of the menu was constant, this leads to spaghetti code and I needed a flexible way to generate the popup menu for the cloud providers plugin

### CODE REVIEW

File: src/View/Sidebar.vala [hunk 1]
```diff
@@ -1834,7 +1813,8 @@ namespace Marlin.Places {
 
                 case Gdk.BUTTON_SECONDARY:
                     if (path != null && !category_at_path (path)) {
-                        popup_menu (event);
+                        //  Add idle to finish the event before open popup menu
+                        Idle.add (() => { show_popup_menu (event); return false; });
```
Comment (jeremypw) [inline src/View/Sidebar.vala]: Indent closure on separate lines

File: src/View/PopupMenuBuilder.vala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    Author(s):  Fernando da Silva Sousa <wild.nando@gmail.com>
+***/
+
+public class PopupMenuBuilder : Object {
+    public delegate void MenuitemCallback (Gtk.MenuItem item);
+    Gtk.MenuItem[] itens = {};
+
+    public Gtk.Menu build () {
+        var popupmenu = new Gtk.Menu ();
+        foreach (var item in itens) {
```
Comment (jeremypw) [inline src/View/PopupMenuBuilder.vala]: Variable name should be `items` or, better, `menu_items`.

File: src/View/Sidebar.vala [hunk 3]
```diff
@@ -1478,80 +1463,79 @@ namespace Marlin.Places {
...
+                        Column.BOOKMARK, out is_bookmark
+            );
+            bool is_plugin = (type == Marlin.PlaceType.PLUGIN_ITEM);
 
-                popupmenu_unmount_item = new Gtk.MenuItem.with_mnemonic (_("_Unmount"));
-                popupmenu_unmount_item.activate.connect (unmount_shortcut_cb);
-                popupmenu_unmount_item.show ();
+            bool show_mount, show_unmount, show_eject, show_rescan, show_format, show_start, show_stop;
+            check_visibility (mount,
+                                volume,
```
Comment (jeremypw) [inline src/View/Sidebar.vala]: Align parameters under the first parameter

File: src/View/Sidebar.vala [hunk 4]
```diff
@@ -1478,80 +1463,79 @@ namespace Marlin.Places {
...
+                var menu = new PopupMenuBuilder ()
+                .add_open (open_shortcut_cb)
+                .add_separator ()
+                .add_open_tab (open_shortcut_in_new_tab_cb)
+                .add_open_window (open_shortcut_in_new_window_cb);
+
+                if (is_bookmark) { menu.add_separator ().add_remove (remove_shortcut_cb).add_rename (rename_shortcut_cb); }
+                if (show_mount) { menu.add_separator ().add_mount (mount_selected_shortcut); }
+                if (show_unmount) { menu.add_separator ().add_unmount (unmount_shortcut_cb); }
+                if (show_eject) { menu.add_separator ().add_eject (eject_shortcut_cb); }
```
Comment (jeremypw) [inline src/View/Sidebar.vala]: Single line if clauses ( x4).  Elementary code style always indents clause on separate lines.
```
if (...) {
    ...;
}
```

File: src/View/Sidebar.vala [hunk 5]
```diff
@@ -1478,80 +1463,79 @@ namespace Marlin.Places {
...
-                popupmenu.append (popupmenu_separator_item1);
-                popupmenu.append (popupmenu_remove_item);
-                popupmenu.append (popupmenu_rename_item);
-                popupmenu.append (popupmenu_separator_item2);
-                popupmenu.append (popupmenu_mount_item);
-                popupmenu.append (popupmenu_unmount_item);
-                popupmenu.append (popupmenu_eject_item);
-                popupmenu.append (popupmenu_empty_trash_item);
-                popupmenu.append (popupmenu_drive_property_item);
+                if (show_property) { menu.add_property (show_drive_info_cb); }
```
Comment (jeremypw) [inline src/View/Sidebar.vala]: Single line if clause

**Comment by naaando** [comment]
Hey @jeremypw, thanks for the amazing review, and about the `if (x) { ...; }` I wrote like this because I already saw it somewhere in elementary's code and thought it was looking good. The behavior bug was also fixed.

**Comment by jeremypw** [comment]
@naaando : Yes, old elementary code often does not fully comply with current coding guidelines. It generally just get fixed when touched or as part of a tidying up only PR.  I was going by the example to the  guideline at https://elementary.io/docs/code/reference#indentation (although the text does not explicitly rule out a single line form - only requiring braces).

**Comment by naaando** [comment]
Fixed @jeremypw ;)

**Pull request merged**